### ISSUE TITLE
Setting custom price on bundle product causes wrong prices on attached simple products when reordering.

### ISSUE THREAD

**Issue by Hexmage** [magento/magento2#30343]
### Preconditions (*)

1. Magento 2.3.5-p2 and 2.4-develop

### Steps to reproduce (*)

1. Create an order containing a bundle product with a fixed price. Give the bundle a custom price.
2. Reorder the previous order

### Expected result (*)

1. The Reordered order is identical to the first order.

### Actual result (*)

1. The reordered order's products in the the bundle all receive the custom price.
![Screenshot from 2020-10-05 12-13-42](https://user-images.githubusercontent.com/22170051/95067542-4c854c80-0704-11eb-9f12-a05387f7cae2.png)
Order 88 is the original order. Order 89 is the reordered order.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### Additional info from Engcom
Steps to reproduce
1. Have a Bundle product
2. From Admin, Create an order adding the Bundle product
3. Set _Custom price_ while performing the order. Submit order
![custom_price1](https://user-images.githubusercontent.com/51680745/96125980-a8b65080-0efc-11eb-8c23-0da6f81c2d04.png)
4. Open `sales_order_item` table and check the correct prices for the product items
5. From Admin panel, **Reorder** the previously created order changing nothing in the order
6. Open `sales_order_item` table and check the prices for the product items

Actual result
The product items prices have been changed 
![custom_price2](https://user-images.githubusercontent.com/51680745/96128005-70634200-0efd-11eb-8523-f3439a139f9d.png)

**Comment by Hexmage** [comment]
When originally ordering the order the custom price is set after the simples are added to the quote. In case of a reorder this happens at the same time causing the custom price to be copied from the bundle to the simples. 

This only causes issues if you use the info of the simple products in an order export. Which I'm doing.

**Comment by Hexmage** [comment]
The same happens when editing an order.

**Comment by Hexmage** [comment]
Turns out it happens with any product that has a parent.

**Comment by Hexmage** [comment]
Potential fix for the issue.

\Magento\Quote\Model\Quote::addProduct

````    /**
     * Add product. Returns error message if product type instance can't prepare product.
     *
     * @param mixed $product
     * @param null|float|\Magento\Framework\DataObject $request
     * @param null|string $processMode
     * @return \Magento\Quote\Model\Quote\Item|string
     * @throws \Magento\Framework\Exception\LocalizedException
     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
     * @SuppressWarnings(PHPMD.NPathComplexity)
     */
    public function addProduct(
        \Magento\Catalog\Model\Product $product,
        $request = null,
        $processMode = \Magento\Catalog\Model\Product\Type\AbstractType::PROCESS_MODE_FULL
    ) {
        if ($request === null) {
            $request = 1;
        }
        if (is_numeric($request)) {
            $request = $this->objectFactory->create(['qty' => $request]);
        }
        if (!$request instanceof \Magento\Framework\DataObject) {
            throw new \Magento\Framework\Exception\LocalizedException(
                __('We found an invalid request for adding product to quote.')
            );
        }

        if (!$product->isSalable()) {
            throw new \Magento\Framework\Exception\LocalizedException(
                __('Product that you are trying to add is not available.')
            );
        }

        $cartCandidates = $product->getTypeInstance()->prepareForCartAdvanced($request, $product, $processMode);

        /**
         * Error message
         */
        if (is_string($cartCandidates) || $cartCandidates instanceof \Magento\Framework\Phrase) {
            return (string)$cartCandidates;
        }

        /**
         * If prepare process return one object
         */
        if (!is_array($cartCandidates)) {
            $cartCandidates = [$cartCandidates];
        }

        $parentItem = null;
        $errors = [];
        $item = null;
        $items = [];
        foreach ($cartCandidates as $candidate) {
 +           $cloneRequest = clone $request;
            // Child items can be sticked together only within their parent
            $stickWithinParent = $candidate->getParentProductId() ? $parentItem : null;
            $candidate->setStickWithinParent($stickWithinParent);

            $item = $this->getItemByProduct($candidate);
            if (!$item) {
+                if($candidate->getParentProductId()) {
+                    $cloneRequest->unsetData('custom_price');
+                }
-               $item = $this->itemProcessor->init($candidate, $request);
+               $item = $this->itemProcessor->init($candidate, $cloneRequest);
                $item->setQuote($this);
                $item->setOptions($candidate->getCustomOptions());
                $item->setProduct($candidate);
                // Add only item that is not in quote already
                $this->addItem($item);
            }
            $items[] = $item;

            /**
             * As parent item we should always use the item of first added product
             */
            if (!$parentItem) {
                $parentItem = $item;
            }
            if ($parentItem && $candidate->getParentProductId() && !$item->getParentItem()) {
                $item->setParentItem($parentItem);
            }

-             $this->itemProcessor->prepare($item, $request, $candidate);
+            $this->itemProcessor->prepare($item, $cloneRequest, $candidate);

            // collect errors instead of throwing first one
            if ($item->getHasError()) {
                foreach ($item->getMessage(false) as $message) {
                    if (!in_array($message, $errors)) {
                        // filter duplicate messages
                        $errors[] = $message;
                    }
                }
            }
        }
        if (!empty($errors)) {
            throw new \Magento\Framework\Exception\LocalizedException(__(implode("\n", $errors)));
        }

        $this->_eventManager->dispatch('sales_quote_product_add_after', ['items' => $items]);
        return $parentItem;
    }````

**Comment by engcom-Bravo** [comment]
Hello @Hexmage 

Thank you for your report.

We have tried to reproduce the issue using the steps provided in the description
When we **Reorder** the previously created order from Admin, and set **Custom price** different from the original 
![update_price_custom](https://user-images.githubusercontent.com/51680745/96119782-262a9280-0ef6-11eb-9246-49f4363191e6.png)
the price of the Bundle item is changed in `sales_order_item` table. For the simple products in the Bundle, the price is not changed
![reorder](https://user-images.githubusercontent.com/51680745/96120261-ea43fd00-0ef6-11eb-953c-067264217f4d.png)
So we can not reproduce the issue with the provided description 

### BUT
When we **Reorder** and add another simple product, and set Custom price for both items,
![update_price_custom](https://user-images.githubusercontent.com/51680745/96121187-5f640200-0ef8-11eb-90ff-80df7e975d1a.png)
the price of the child products is also changed
![reorder3](https://user-images.githubusercontent.com/51680745/96121257-7e629400-0ef8-11eb-8e37-321c1db9bd2a.png)

So my question is, do you add another product while Reordering?

If no, can you please provide more detailed **Steps to reproduce**? In this way, you will help us to proceed faster with the issue

Thank you in advance

**Comment by Hexmage** [comment]
@engcom-Bravo 

Step 1 Create an order.
Step 2 Add bundle product (or configurable product because the same thing happens)
Step 3 Set a custom price on the bundle
Step 4 Finish order
Step 5 Reorder previous Order
Step 6 Change nothing on the reordered order and finish it

**Comment by engcom-Bravo** [comment]
Thank you @Hexmage for the quick response. 
We confirm that the issue is reproducible on 2.4-develop, and we have added the provided info to the Description

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-38460` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
Hi @Hexmage. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-38460` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.2` release.
